cas.o: Keep unreferenced symbols alive by default

dexonsmith opened the pull request:
Stop dropping unreferenced symbols by default. Store them in an extra
table in the compile unit. Among other things, this keeps debug info
alive by default.

There are a few new options:

- `--dead-strip` restores the old behaviour.
- `--dead-strip-section` just dead-strips the named section. Can be
  specified multiple times.
- `--keep-alive-section` keeps the named section alive, adding the
  "no-dead-strip" attribute to any symbols in it. It's stronger than
  the `--dead-strip*` options.

dexonsmith committed:
cas.o: Keep unreferenced symbols alive by default

Stop dropping unreferenced symbols by default. Store them in an extra
table in the compile unit. Among other things, this keeps debug info
alive by default.

There are a few new options:

- `--dead-strip` restores the old behaviour.
- `--dead-strip-section` just dead-strips the named section. Can be
  specified multiple times.
- `--keep-alive-section` keeps the named section alive, adding the
  "no-dead-strip" attribute to any symbols in it. It's stronger than
  the `--dead-strip*` options.

dexonsmith commented:
@swift-ci test

dexonsmith commented:
Really there should be tests for these, but:
- `llvm-cas-object-format` currently lacks output good for testing this,
- these are `cl::opt`s defined as `static`s so compiled unit tests won't work, and
- threading the options through the `CompileUnitRef::create` APIs is a lot bigger than just this piece.

@akyrtzi  @cachemeifyoucan , are you comfortable landing as-is (once tests pass)?

cachemeifyoucan commented:
I think there should be a test for the default behavior at least.

`--dump` option for llvm-cas-object-format can at least show an abstract entry for the symbol right or is the format to obscure to test with?

akyrtzi commented:
> llvm-cas-object-format currently lacks output good for testing this,

Would `--print-cas-object` help with this?

dexonsmith committed:
cas.o: Add tests for keeping unreferenced symbols alive

Add tests, and fix the following:

- Fix `--{dead-strip,keep-alive}-section`, which didn't work before
  because they used `cl::opt<std::vector<std::string>>` instead of
  `cl::list<std::string>`.
- Add `--print-cas-object-sort-sections` to make section order
  easily predictable (regardless of what order the CAS reports symbols
  in).
- Change `casobjectformats::printCASObject`'s output to be easier to
  FileCheck by replacing:
        Symbols:
          _sym1 ...
          _sym2 ...
  with:
        SYMBOL: _sym1 ...
        SYMBOL: _sym2 ...
  (and a similar change for fixups, although I didn't need that here).

dexonsmith commented:
@swift-ci test

dexonsmith commented:
> Would `--print-cas-object` help with this?
Yes!

> I think there should be a test for the default behavior at least.
Thanks both for not letting me off the hook!

Tests added.

dexonsmith commented:
@swift-ci macos test

dexonsmith commented:
@swift-ci test macos

dexonsmith commented:
@swift-ci test

dexonsmith commented:
@swift-ci please test?

dexonsmith committed:
cas.o: Fix llvm/test/tools/llvm-cas-object-format/print-object.test

dexonsmith commented:
Ran tests locally. Pushed commit to fix test for CAS object output.
